x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -421,6 +437,62 @@ public Entry<CacheFileRegion> get(KeyType cacheKey, long fileLength, int region)
         return entry;
     }
 
+    /**
+     * Fetch and cache the full blob for the given cache entry from the remote repository if there
+     * are enough free pages in the cache to do so.
+     *
+     * This method returns as soon as the download tasks are instantiated, but the tasks themselves
+     * are run on the bulk executor.
+     *
+     * If an exception is thrown from the writer then the cache entry being downloaded is freed
+     * and unlinked
+     *
+     * @param cacheKey  the key to fetch data for
+     * @param length    the length of the blob to fetch
+     * @param writer    a writer that handles writing of newly downloaded data to the shared cache
+     *
+     * @return {@code true} if there were enough free pages to start downloading
+     */
+    public boolean maybeFetchFullEntry(KeyType cacheKey, long length, RangeMissingHandler writer) {
+        int finalRegion = getEndingRegion(length);
+        if (freeRegionCount() < finalRegion) {
+            // Not enough room to download a full file without evicting existing data, so abort
+            return false;
+        }
+        long regionLength = regionSize;
+        for (int region = 0; region <= finalRegion; region++) {
+            var entry = get(cacheKey, length, region);
+            if (region == finalRegion) {
+                regionLength = length - getRegionStart(region);
+            }
+            ByteRange rangeToWrite = ByteRange.of(0, regionLength);
+            if (rangeToWrite.length() == 0) {
+                return true;
+            }
+            entry.chunk.populateAndRead(
+                rangeToWrite,
+                ByteRange.EMPTY,
+                (channel, pos, relativePos, len) -> 0,
+                writer,
+                bulkIOExecutor,
+                ActionListener.noop(),
+                e -> {
+                    // if an exception is thrown during download then evict the entry
```

Comment:
Why do you think we need this? Isn't this a rather uncommon case, it really really should be for our assumptions elsewhere to hold?

Reply:
This was how I was planning on implementing cancellation - the write handler checks a flag when it runs to see if it still needs to read data, and if it doesn't then it throws and exception which this check will catch and evict the entry.  Do you have a different way of doing things in mind?